### ISSUE TITLE
cannot find matching URI scheme

### ISSUE THREAD

**Issue by christian-rauch** [stevenlovegrove/Pangolin#669]
Since the component refactoring, I am not able to start the example programs on Linux/X11:
```
$ ./examples/SimpleDisplay/SimpleDisplay 
terminate called after throwing an instance of 'pangolin::FactoryRegistry::NoMatchingSchemeException'
  what():  Unable to open URI . No matching scheme.
Aborted (core dumped)
```

**Comment by christian-rauch** [comment]
@stevenlovegrove  The commit that broke this is: c3e96c8a39c05e52bee9537b122128993e3ba0e6

**Comment by stevenlovegrove** [comment]
Sorry about that! Thanks for reporting.

I'm not sure if I should put those files back using proper build rules this time, or try to get a cross platform plugin system working. It's currently relying on static initialization but Linux must be stripping those symbols or something.

Is it dynamic or static linking the components?

**Comment by christian-rauch** [comment]
> Is it dynamic or static linking the components?

I don't know :-) This happens with the default settings (cmake .. && make).

> It's currently relying on static initialization but Linux must be stripping those symbols or something.

Is this about the symbol visibility? I think the windowing systems are statically linked, so the symbols will be removed. You can adjust the visibility of symbols in GCC: https://gcc.gnu.org/onlinedocs/gcc-4.1.2/gcc/Function-Attributes.html.

Which are the plugins in Pangolin? Are you trying to separate the different windowing systems into separate plugin files (so) that are loaded dynamically? According to lddtree:
```
libpango_windowing.so => /home/christian/Development/Pangolin/build/libpango_windowing.so (interpreter => none)
    libX11.so.6 => /lib/x86_64-linux-gnu/libX11.so.6
        libxcb.so.1 => /lib/x86_64-linux-gnu/libxcb.so.1
            libXau.so.6 => /lib/x86_64-linux-gnu/libXau.so.6
            libXdmcp.so.6 => /lib/x86_64-linux-gnu/libXdmcp.so.6
                libbsd.so.0 => /lib/x86_64-linux-gnu/libbsd.so.0
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2
            ld-linux-x86-64.so.2 => /lib/x86_64-linux-gnu/ld-linux-x86-64.so.2
    libEGL.so.1 => /lib/x86_64-linux-gnu/libEGL.so.1
        libGLdispatch.so.0 => /lib/x86_64-linux-gnu/libGLdispatch.so.0
    libpango_core.so => /home/christian/Development/Pangolin/build/libpango_core.so
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0
    libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6
    libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1
    libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6
```
the whole windowing system is in a single shared library and I don't see other references to `dlopen` in the source.

**Comment by stevenlovegrove** [comment]
https://github.com/stevenlovegrove/Pangolin/blob/master/components/pango_core/include/pangolin/factory/factory_registry.h#L192 is the static initializer doing the work. It should get invoked as soon as the compile unit is loaded. I had assumed this would happen without issue for video and windowing because you have to call into that compile unit with the Open method.

It gets more difficult with the Python interpreter which is decoupled from the Console in the display library. With both Windowing and Video, the idea is that anyone can add their own through external libs. I don't yet have any dlopen calls, but I was considering adding it.

Do you think that macro can be modified to force inclusion of the symbols on Linux so that they get loaded, at least during the Open call into the respective libraries (pango_video and pango_windowing)?

**Comment by stevenlovegrove** [comment]
And here https://github.com/stevenlovegrove/Pangolin/blob/master/components/pango_windowing/src/display_x11.cpp#L507 is use of that macro for linux, btw.

**Comment by stevenlovegrove** [comment]
Okay, this is now working again (as of 2e74d1f9286d466d467bf176a69153711076dd75) with a simple enough fix since pango_windowing is an explicit dependency of pango_display anyway. Tested with Ubuntu 21.04

The video URI loading wasn't broken because symbols from pango_video were already getting touched when attempting to load any driver (which I copied for windowing), but there remains an issue with the python interpreter which is completely decoupled from the console in pango_display. I'll need to do some .so searching in the future, but in the meantime the workaround for the python console in Linux is to explicitly preload the library:
`LD_PRELOAD=libpango_python.so ./examples/SimpleDisplay/SimpleDisplay`
And then hit \` to bring up the console.

I just checked, and you also will have problems when building pangolin statically (i.e. with `cmake -DBUILD_SHARED_LIBS=OFF ..`) so there is still some work to do here...

**Comment by christian-rauch** [comment]
I don't really understand what the underlying problem is. Is there a public symbol missing from one of the so files?

The plugin systems that I have used so far load plugins (as so files) via `dlopen` and then search for an entry symbol/point via `dlsym`. In a shared library, the (public) symbols must be exported. As far as I know, this only applies to non-static functions. I.e. static functions in C will not export their symbols and the function name will not be visible when listing the symbols via `nm -C <library>.so`. I think this is the same for static libraries.

**Comment by christian-rauch** [comment]
I still have this issue for the headless backend, e.g. with the `HelloPangolinOffscreen` example.

The primary reason here is that the factory is not registered. But even after adding:
```
PangolinRegisterFactory(WindowInterface HeadlessWindow)
```
to the CMakeLists, and using the macro like:
```
PANGOLIN_REGISTER_FACTORY(HeadlessWindow) {
struct HeadlessWindowFactory : public TypedFactoryInterface<WindowInterface> {
```
it will still give me: `Unable to open URI . Unrecognized options for scheme (window_title,scheme,)`.

**Comment by stevenlovegrove** [comment]
@christian-rauch aha, you hit Pangolin's new parameter checking! The HeadlessWindowFactory was being passed window_title and scheme which it wasn't expecting. I just pushed a fix 73b9586d290a6f81117af66ce69ca070ae522de5

**Comment by stevenlovegrove** [comment]
> I don't really understand what the underlying problem is. Is there a public symbol missing from one of the so files?

Don't worry about this - I'm on top of the problem.

**Comment by christian-rauch** [comment]
The `HelloPangolinOffscreen` example worked (it was able to create the context) with your fix in 73b9586d290a6f81117af66ce69ca070ae522de5 but it now broke again in the very next commit a0d0b5173c74906a0ff96f654fea7616755bc6f4 :-)

**Comment by stevenlovegrove** [comment]
haha, thanks! Fixed here: 45cc981dab61d5bdfc35890a8797975b39e02187

I reverted to the file based registration for built-in drivers again because it is just way more reliable.
